jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererCodeCoverageTracker.kt

```diff
@@ -0,0 +1,143 @@
+                override fun afterAccept(states: InvocationContext, sessionContext: SessionContext, remainingRecomm: String) {
+                    if (states.requestContext.fileContextInfo.programmingLanguage.toCodeWhispererLanguage() != language) return
+                    addAndGetAcceptedTokens(remainingRecomm.length)
+                }
+            }
+        )
+        scheduleCodeWhispererCodeCoverageTracker()
+    }
+
+    fun documentChanged(event: DocumentEvent) {
+        addAndGetTotalTokens(event.newLength - event.oldLength)
+    }
+
+    private fun flush() {
+        try {
+            if (isTelemetryEnabled()) emitCodeWhispererCodeContribution()
+        } finally {
+            scheduleCodeWhispererCodeCoverageTracker()
+            init()
+        }
+    }
+
+    private fun scheduleCodeWhispererCodeCoverageTracker() {
+        if (!alarm.isDisposed && !isShuttingDown.get()) {
+            alarm.addRequest({ flush() }, Duration.ofSeconds(timeWindowInSec).toMillis())
+        }
+    }
+
+    private fun addAndGetAcceptedTokens(delta: Int): Int =
+        if (!isTelemetryEnabled()) acceptedTokensSize.get()
+        else acceptedTokensSize.addAndGet(delta)
+
+    private fun addAndGetTotalTokens(delta: Int): Int =
+        if (!isTelemetryEnabled()) totalTokensSize.get()
+        else {
+            val result = totalTokensSize.addAndGet(delta)
+            if (result < 0) totalTokensSize.set(0)
+            result
+        }
+
+    private fun init() {
+        startTime = Instant.now()
+        totalTokensSize.set(0)
+        acceptedTokensSize.set(0)
+    }
+
+    private fun emitCodeWhispererCodeContribution() {
+        CodewhispererTelemetry.codePercentage(
+            project = null,
+            acceptedTokensSize.get(),
+            language,
+            percentage,
+            startTime.toString(),
+            totalTokensSize.get()
+        )
+    }
+
+    @TestOnly
+    fun forceTrackerFlush() {
+        alarm.drainRequestsInTest()
+    }
+
+    @TestOnly
+    fun activeRequestCount() = alarm.activeRequestCount
+
+    override fun dispose() {
+        if (isShuttingDown.getAndSet(true)) {
+            return
+        }
+        flush()
+    }
+
+    companion object {
+        private val instances: MutableMap<CodewhispererLanguage, CodeWhispererCodeCoverageTracker> = mutableMapOf()
+
+        fun getInstance(language: CodewhispererLanguage): CodeWhispererCodeCoverageTracker = when (val instance = instances[language]) {
+            null -> {
+                val newTracker = DefaultCodeWhispererCodeCoverageTracker(language)
+                instances[language] = newTracker
+                newTracker
+            }
+            else -> instance
+        }
+
+        @TestOnly
+        fun getInstancesMap(): MutableMap<CodewhispererLanguage, CodeWhispererCodeCoverageTracker> {
+            assert(ApplicationManager.getApplication().isUnitTestMode)
+            return instances
+        }
```

Comment:
nit: Or you can just do `private set` on `val instances`

Reply:
but Abhilash, with `val instances: MutableMap<T,E>` , we already have no setter on this static field. And for testing, we only need its getter, so the easiest way I think it to provide an api to get this map but restricted to unittest mode.

Reply:
That was my miss. nvm.